README.md
# dart-rencode

Rencoding implementation in dart.

Rencoding is a more complicated version of [bencoding](https://en.wikipedia.org/wiki/Bencode) used by the Deluge torrent client and its RPC APIS.

It produces smaller encoding for small inputs than the original bencode.

**Note**

You probably don't need this unless you are writing a Deluge RPC client.


CHANGELOG.md
## [0.0.1] - 2018-03-29

* Initial release

## [0.0.2] - 2018-06-24

* Dart 2 fixes.

## [0.0.3] - 2018-09-22

* Updated dependencies

## [0.0.4] - 2021-03-17

* Update dependencies, sdk constraints

## [0.0.5] - 2021-03-18

* Null safety


test/rencode_test.dart
import 'dart:math';

import 'package:test/test.dart';

import 'package:rencode/rencode.dart';

void main() {
  test('Test encoding', () {
    //All the hardcoded arrays are output of original python rencode implementation

    var codec = RencodeCodec();

    //Positive embedded ints
    expect(codec.encode(0), [0]);
    expect(codec.encode(43), [43]);

    //Negative embedded ints
    expect(codec.encode(-1), [70]);
    expect(codec.encode(-32), [101]);

    //Bytes
    expect(codec.encode(44), [62, 44]);
    expect(codec.encode(127), [62, 127]);
    expect(codec.encode(-33), [62, 223]);
    expect(codec.encode(-128), [62, 128]);

    //Shorts
    expect(codec.encode(128), [63, 0, 128]);
    expect(codec.encode(32767), [63, 127, 255]);
    expect(codec.encode(-129), [63, 255, 127]);
    expect(codec.encode(-32768), [63, 128, 0]);

    //Ints
    expect(codec.encode(32768), [64, 0, 0, 128, 0]);
    expect(codec.encode(2147483647), [64, 127, 255, 255, 255]);
    expect(codec.encode(-32769), [64, 255, 255, 127, 255]);
    expect(codec.encode(-2147483648), [64, 128, 0, 0, 0]);

    //Longs
    expect(codec.encode(2147483648), [65, 0, 0, 0, 0, 128, 0, 0, 0]);
    expect(codec.encode(9223372036854775807), [65, 127, 255, 255, 255, 255, 255, 255, 255]);
    expect(codec.encode(-2147483649), [65, 255, 255, 255, 255, 127, 255, 255, 255]);
    expect(codec.encode(-9223372036854775808), [65, 128, 0, 0, 0, 0, 0, 0, 0]);

    //Numbers
    expect(codec.encode(BigInt.parse('9223372036854775808')), [NUMBER, ...'9223372036854775808'.codeUnits, END]);

    //Null
    expect(codec.encode(null), [NULL]);

    //Bools
    expect(codec.encode(true), [TRUE]);
    expect(codec.encode(false), [FALSE]);

    //Fixed map
    expect(codec.encode({'a' : 'b'}), [103, 129, 97, 129, 98]);
    expect(codec.encode({'a' : 1, 'b' : 2, '3' : 3}), [105, 129, 97, 1, 129, 98, 2, 129, 51, 3]);

    //Fixed list
    expect(codec.encode(['a','b','b']), [195, 129, 97, 129, 98, 129, 98]);
    expect(codec.encode([1,2,3,4]), [196, 1, 2, 3, 4]);

    //Strings
    expect(codec.encode('abcdefgh'), [136, 97, 98, 99, 100, 101, 102, 103, 104]);
    expect(codec.decode([57, 49, 58, 97, 98, 99, 100, 101, 102, 103, 104, 105, 106, 107, 108, 109, 110,
    111, 112, 113, 114, 115, 116, 117, 118, 119, 120, 121, 122, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57,
    48, 33, 64, 35, 36, 37, 94, 38, 42, 40, 41, 95, 43, 45, 61, 65, 66, 67, 68, 69, 70, 71, 72, 73, 74,
    75, 76, 77, 78, 79, 80, 81, 82, 83, 84, 85, 86, 87, 88, 89, 90, 123, 125, 91, 93, 58, 59, 34, 39,
    44, 46, 60, 62, 63, 47]), "abcdefghijklmnopqrstuvwxyz01234567890!@#\$%^&*()_+-=ABCDEFGHIJKLMNOPQRSTUVWXYZ{}[]:;\"',.<>?/");

    //Random stuff
    expect(codec.encode({'a': 1, 'b': '12', 'c': null, 'd': true, 'e': false, 'f': [1, 2, 3]}), [108, 129, 97, 1, 129, 98, 130, 49, 50, 129, 99, 69, 129, 100, 67, 129, 101, 68, 129, 102, 195, 1, 2, 3]);
    expect(codec.encode({'a': [1, 2, 3, {'b': 'c'}]}), [103, 129, 97, 196, 1, 2, 3, 103, 129, 98, 129, 99]);

    //Recursive list
    expect(codec.encode([[[[[[1], [2], [3]], 4], 5], 6], 7]), [194, 194, 194, 194, 195, 193, 1, 193, 2, 193, 3, 4, 5, 6, 7]);

    //Recursive map
    expect(codec.encode({'a': {'b': {'c': {'d': 'e', 'f': 'g'}, 'h': {'i': 'j'}}}}), [103, 129, 97, 103, 129, 98, 104, 129, 99, 104, 129, 100, 129, 101, 129, 102, 129, 103, 129, 104, 103, 129, 105, 129, 106]);
  });

  test('Test decoding', () {
    //All the hardcoded arrays are output of original python rencode implementation

    var codec = RencodeCodec();

    //Positive embedded ints
    expect(codec.decode([0]), 0);
    expect(codec.decode([43]), 43);

    //Negative embedded ints
    expect(codec.decode([70]), -1);
    expect(codec.decode([101]), -32);

    //Bytes
    expect(codec.decode([62, 44]), 44);
    expect(codec.decode([62, 127]), 127);
    expect(codec.decode([62, 223]), -33);
    expect(codec.decode([62, 128]), -128);

    //Shorts
    expect(codec.decode([63, 0, 128]), 128);
    expect(codec.decode([63, 127, 255]), 32767);
    expect(codec.decode([63, 255, 127]), -129);
    expect(codec.decode([63, 128, 0]), -32768);

    //Ints
    expect(codec.decode([64, 0, 0, 128, 0]), 32768);
    expect(codec.decode([64, 127, 255, 255, 255]), 2147483647);
    expect(codec.decode([64, 255, 255, 127, 255]), -32769);
    expect(codec.decode([64, 128, 0, 0, 0]), -2147483648);

    //Longs
    expect(codec.decode([65, 0, 0, 0, 0, 128, 0, 0, 0]), 2147483648);
    expect(codec.decode([65, 127, 255, 255, 255, 255, 255, 255, 255]), 9223372036854775807);
    expect(codec.decode([65, 255, 255, 255, 255, 127, 255, 255, 255]), -2147483649);
    expect(codec.decode([65, 128, 0, 0, 0, 0, 0, 0, 0]), -9223372036854775808);

    //Numbers
    expect(codec.decode([NUMBER, ...'9223372036854775808'.codeUnits, END]), BigInt.parse('9223372036854775808'));
    expect(codec.decode([NUMBER, ...'9.223372036854775808'.codeUnits, END]), 9.223372036854775808);

    //Null
    expect(codec.decode([NULL]), null);

    //Bools
    expect(codec.decode([TRUE]), true);
    expect(codec.decode([FALSE]), false);

    //Fixed map
    expect(codec.decode([103, 129, 97, 129, 98]), {'a' : 'b'});
    expect(codec.decode([105, 129, 97, 1, 129, 98, 2, 129, 51, 3]), {'a' : 1, 'b' : 2, '3' : 3});

    //Fixed list
    expect(codec.decode([195, 129, 97, 129, 98, 129, 98]), ['a','b','b']);
    expect(codec.decode([196, 1, 2, 3, 4]), [1,2,3,4]);

    //Strings
    expect(codec.decode([136, 97, 98, 99, 100, 101, 102, 103, 104]), 'abcdefgh');
    expect(codec.decode([57, 49, 58, 97, 98, 99, 100, 101, 102, 103, 104, 105, 106, 107, 108, 109, 110,
    111, 112, 113, 114, 115, 116, 117, 118, 119, 120, 121, 122, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57,
    48, 33, 64, 35, 36, 37, 94, 38, 42, 40, 41, 95, 43, 45, 61, 65, 66, 67, 68, 69, 70, 71, 72, 73, 74,
    75, 76, 77, 78, 79, 80, 81, 82, 83, 84, 85, 86, 87, 88, 89, 90, 123, 125, 91, 93, 58, 59, 34, 39,
    44, 46, 60, 62, 63, 47]), "abcdefghijklmnopqrstuvwxyz01234567890!@#\$%^&*()_+-=ABCDEFGHIJKLMNOPQRSTUVWXYZ{}[]:;\"',.<>?/");

    //Random stuff
    expect(codec.decode([108, 129, 97, 1, 129, 98, 130, 49, 50, 129, 99, 69, 129, 100, 67, 129, 101, 68, 129, 102, 195, 1, 2, 3]), {'a': 1, 'b': '12', 'c': null, 'd': true, 'e': false, 'f': [1, 2, 3]});
    expect(codec.decode([103, 129, 97, 196, 1, 2, 3, 103, 129, 98, 129, 99]), {'a': [1, 2, 3, {'b': 'c'}]});

    //Recursive list
    expect(codec.decode([194, 194, 194, 194, 195, 193, 1, 193, 2, 193, 3, 4, 5, 6, 7]), [[[[[[1], [2], [3]], 4], 5], 6], 7]);

    //Recursive map
    expect(codec.decode([103, 129, 97, 103, 129, 98, 104, 129, 99, 104, 129, 100, 129, 101, 129, 102, 129, 103, 129, 104, 103, 129, 105, 129, 106]), {'a': {'b': {'c': {'d': 'e', 'f': 'g'}, 'h': {'i': 'j'}}}});
    });

  test('Test both', () {
    var codec = RencodeCodec();

    expect(codec.decode(codec.encode(1.2345)), 1.2345);
    expect(codec.decode(codec.encode(1234567890123456789)), 1234567890123456789);
    expect(codec.decode(codec.encode(1234567890123456789.1234567890)), 1234567890123456789.1234567890);

    Object ld = [[{'a' : 15, 'bb' : 2.5, 'ccc' : 29.3, '' : [-0.3, <int>[], false, true, '']},
        ['a', 10e20], ...Iterable.generate(100000),], pow(2, 30), pow(2, 33), pow(2, 60), pow(2, 66),
        'b' * 30, 'v' * 33, 'c' * 64, pow(2, 1000), false, true, false, -1, 0, 1];

    expect(codec.decode(codec.encode(ld)), ld);

    ld = ['', 'a'*10, 'b'*100, 'c'*1000, 'd'*10000, 'd'*(10^5), 'e'*(10^6), 'f'*(10^7)];

    expect(codec.decode(codec.encode(ld)), ld);

    expect(codec.decode(codec.encode('ಠ_ಠ')), 'ಠ_ಠ');
  });
}

lib/rencode.dart
library rencode;

import 'dart:convert';
import 'dart:typed_data';
import 'dart:collection';

/*
* The bencode 'typecodes' such as i, d, etc have been extended and
* relocated on the base-256 character set.
*/
const int LIST = 59;
const int DICTIONARY = 60;
const int NUMBER = 61;
const int BYTE = 62;
const int SHORT = 63;
const int INT = 64;
const int LONG = 65;
const int FLOAT = 66;
const int DOUBLE = 44;
const int TRUE = 67;
const int FALSE = 68;
const int NULL = 69;
const int END = 127;
const int LENGTH_DELIM = 58;

//Positive integers with value embedded in typecode.
const int INT_POS_START = 0;
const int INT_POS_COUNT = 44;

//Negative integers with value embedded in typecode.
const int INT_NEG_START = 70;
const int INT_NEG_COUNT = 32;

//Dictionaries with length embedded in typecode.
const int DICT_START = 102;
const int DICT_COUNT = 25;

//Strings with length embedded in typecode.
const int STR_START = 128;
const int STR_COUNT = 64;

//Lists with length embedded in typecode.
const int LIST_START = STR_START + STR_COUNT;
const int LIST_COUNT = 64;

class RencodeCodec extends Codec<Object?, List<int>> {
  @override
  Encoder get encoder {
    return Encoder();
  }

  @override
  Decoder get decoder {
    return Decoder();
  }
}

class Encoder extends Converter<Object?, List<int>> {
  late Queue<int> _output;

  @override
  List<int> convert(Object? input) {
    _output = ListQueue();
    _writeObject(input);
    return _output.toList();
  }

  void _writeObject(Object? object) {
    if (object is Map) {
      _writeMap(object as Map<Object, Object?>);
    } else if (object is Iterable) {
      _writeList(object as List<Object?>);
    } else if (object is int) {
      _writeInt(object);
    } else if (object is double) {
      _writeDouble(object);
    } else if (object is BigInt) {
      _writeBigint(object);
    } else if (object is bool) {
      _writeBool(object);
    } else if (object == null) {
      _writeNull();
    } else if (object is String) {
      _writeString(object);
    } else {
      throw ArgumentError('Object of type ${object.runtimeType.toString()} is not supported');
    }
  }

  void _writeMap(Map<Object, Object?> map) {
    if (map.length < DICT_COUNT) {
      _output.add(DICT_START + map.length);
      map.forEach((k, v) {
        _writeObject(k);
        _writeObject(v);
      });
    } else {
      _output.add(DICTIONARY);
      map.forEach((k, v) {
        _writeObject(k);
        _writeObject(v);
      });
      _output.add(END);
    }
  }

  void _writeList(Iterable<Object?> iterable) {
    var length = iterable.length;
    if (length < LIST_COUNT) {
      _output.add(LIST_START + length);
      iterable.forEach((o) => _writeObject(o));
    } else {
      _output.add(LIST);
      iterable.forEach((o) => _writeObject(o));
      _output.add(END);
    }
  }

  void _writeInt(int i) {
    if (i >= 0 && i < INT_POS_COUNT) {
      _output.add(INT_POS_START + i);
    } else if (i >= -INT_NEG_COUNT && i < 0) {
      _output.add(INT_NEG_START - 1 - i);
    } else if (i >= -128 && i < 128) {
      var b = ByteData(1);
      b.setUint8(0, i);
      _output.add(BYTE);
      _output.addAll(b.buffer.asUint8List());
    } else if (i >= -32768 && i < 32768) {
      var b = ByteData(2);
      b.setUint16(0, i);
      _output.add(SHORT);
      _output.addAll(b.buffer.asUint8List());
    } else if (i >= -2147483648 && i < 2147483648) {
      var b = ByteData(4);
      b.setUint32(0, i);
      _output.add(INT);
      _output.addAll(b.buffer.asUint8List());
    } else if (i >= -9223372036854775808 && i <= 9223372036854775807) {
      var b = ByteData(8);
      b.setUint64(0, i);
      _output.add(LONG);
      _output.addAll(b.buffer.asUint8List());
    } else {
      _writeBigint(BigInt.from(i));
    }
  }

  void _writeDouble(double d) {
    var b = ByteData(8);
    b.setFloat64(0, d);
    _output.add(DOUBLE);
    _output.addAll(b.buffer.asUint8List());
  }

  void _writeBigint(BigInt b) {
    _output.add(NUMBER);
    _output.addAll(utf8.encode(b.toString()));
    _output.add(END);
  }

  void _writeBool(bool b) {
    _output.add(b ? TRUE : FALSE);
  }

  void _writeNull() {
    _output.add(NULL);
  }

  void _writeString(String s) {
    var utf8String = utf8.encode(s);
    if (utf8String.length < STR_COUNT) {
      _output.add(STR_START + utf8String.length);
      _output.addAll(utf8String);
    } else {
      var lengthStr = utf8String.length.toString();
      _output.addAll(utf8.encode(lengthStr));
      _output.add(LENGTH_DELIM);
      _output.addAll(utf8String);
    }
  }
}

class Decoder extends Converter<List<int>, Object?> {
  late Queue<int> _input;

  @override
  Object? convert(List<int> input) {
    _input = ListQueue.from(input);
    try {
      return _readObject();
    } catch (e) {
      if (e is FormatException) {
        rethrow;
      } else if (e is StateError) {
        throw FormatException('Malformed rencode ', input);
      } else {
        rethrow;
      }
    }
  }

  void _removeMany(int n) {
    for (var i = 0; i < n; i++) {
      _input.removeFirst();
    }
  }

  Object? _readObject() {
    var token = _input.first;
    if (isMap(token) || isFixedMap(token)) {
      return _readMap();
    } else if (isList(token) || isFixedList(token)) {
      return _readList();
    } else if (isInt(token) || isEmbeddedPositiveInt(token) || isEmbeddedNegativeInt(token)) {
      return _readInt();
    } else if (isDouble(token)) {
      return _readDouble();
    } else if (isNumber(token)) {
      return _readNumber();
    } else if (isBool(token)) {
      return _readBool();
    } else if (token == NULL) {
      _input.removeFirst();
      return null;
    } else {
      return _readStringOrBytes();
    }
  }

  bool isMap(int typeCode) {
    return typeCode == DICTIONARY;
  }

  bool isFixedMap(int typeCode) {
    return typeCode >= DICT_START && typeCode < (DICT_START + DICT_COUNT);
  }

  bool isList(int typeCode) {
    return typeCode == LIST;
  }

  bool isFixedList(int typeCode) {
    return typeCode >= LIST_START && typeCode < (LIST_START + LIST_COUNT);
  }

  bool isInt(int typeCode) {
    return typeCode == BYTE || typeCode == SHORT || typeCode == INT || typeCode == LONG;
  }

  bool isEmbeddedPositiveInt(int typeCode) {
    return typeCode >= INT_POS_START && typeCode < (INT_POS_START + INT_POS_COUNT);
  }

  bool isEmbeddedNegativeInt(int typeCode) {
    return typeCode >= INT_NEG_START && typeCode < (INT_NEG_START + INT_NEG_COUNT);
  }

  bool isDouble(int typeCode) {
    return typeCode == FLOAT || typeCode == DOUBLE;
  }

  bool isNumber(int typeCode) {
    return typeCode == NUMBER;
  }

  bool isBool(int typeCode) {
    return typeCode == TRUE || typeCode == FALSE;
  }

  bool isFixedString(int typeCode) {
    return typeCode >= STR_START && typeCode < (STR_START + STR_COUNT);
  }

  Map<Object, Object?> _readMap() {
    var token = _input.removeFirst();

    var map = <Object, Object?>{};
    if (isFixedMap(token)) {
      var length = token - DICT_START;
      for (var i = 0; i < length; i++) {
        var key = _readObject()!;
        var value = _readObject();
        map[key] = value;
      }
    } else {
      while (_input.first != END) {
        var key = _readObject()!;
        var value = _readObject();
        map[key] = value;
      }
      _input.removeFirst();
    }

    return map;
  }

  List<Object?> _readList() {
    var token = _input.removeFirst();

    var list = <Object?>[];
    if (isFixedList(token)) {
      var length = token - LIST_START;
      for (var i = 0; i < length; i++) {
        list.add(_readObject());
      }
    } else {
      while (_input.first != END) {
        list.add(_readObject());
      }
      _input.removeFirst();
    }

    return list;
  }

  int _readInt() {
    var token = _input.removeFirst();

    if (isEmbeddedPositiveInt(token)) {
      return INT_POS_START + token;
    } else if (isEmbeddedNegativeInt(token)) {
      return INT_NEG_START - 1 - token;
    } else {
      var i = 0;
      if (token == BYTE) {
        var list = Uint8List.fromList(_input.take(1).toList());
        i = list.buffer.asByteData().getInt8(0);
        _removeMany(1);
      } else if (token == SHORT) {
        var list = Uint8List.fromList(_input.take(2).toList());
        i = list.buffer.asByteData().getInt16(0);
        _removeMany(2);
      } else if (token == INT) {
        var list = Uint8List.fromList(_input.take(4).toList());
        i = list.buffer.asByteData().getInt32(0);
        _removeMany(4);
      } else if (token == LONG) {
        var list = Uint8List.fromList(_input.take(8).toList());
        i = list.buffer.asByteData().getInt64(0);
        _removeMany(8);
      }
      return i;
    }
  }

  double _readDouble() {
    var token = _input.removeFirst();

    var d = 0.0;
    if (token == FLOAT) {
      var list = Uint8List.fromList(_input.take(4).toList());
      d = list.buffer.asByteData().getFloat32(0);
      _removeMany(4);
    } else if (token == DOUBLE) {
      var list = Uint8List.fromList(_input.take(8).toList());
      d = list.buffer.asByteData().getFloat64(0);
      _removeMany(8);
    }
    return d;
  }

  Object _readNumber() {
    _input.removeFirst();

    var num = _input.takeWhile((i) => i != END).toList();
    var numStr = utf8.decode(num);

    _removeMany(num.length + 1);

    if (numStr.contains('.')) {
      return double.parse(numStr);
    } else {
      return BigInt.parse(numStr);
    }
  }

  bool _readBool() {
    var token = _input.removeFirst();
    if (token == TRUE) {
      return true;
    } else {
      return false;
    }
  }

  Object _readStringOrBytes() {
    var bytes = _readByteString();
    try {
      return utf8.decode(bytes);
    } on FormatException {
      return bytes;
    }
  }

  List<int> _readByteString() {
    var token = _input.first;

    if (isFixedString(token)) {
      _input.removeFirst();
      var length = token - STR_START;
      var bytes = _input.take(length).toList();
      _removeMany(length);
      return bytes;

    } else if (token >= '1'.codeUnitAt(0) && token <= '9'.codeUnitAt(0)) {
      var length = _input.takeWhile((i) => i != LENGTH_DELIM).toList();
      var lengthStr = utf8.decode(length);
      var lengthInt = int.parse(lengthStr);

      _removeMany(length.length + 1);

      var bytes = _input.take(lengthInt).toList();
      _removeMany(lengthInt);
      return bytes;
    } else {
      throw FormatException('Malformed rencode ', _input);
    }
  }
}